PROTOCOL.md

```diff
@@ -0,0 +1,74 @@
+# GraphQL over WebSocket Protocol
+
+## Client-server communication
+
+Each message has a `type` field, which defined in the protocol of this package, as well as associated fields inside `payload` field, depending on the message type.
+
+### Client -> Server
+
+#### GQL_CONNECTION_INIT
+Client sends this message after connecting, this triggers `onConnect` on the server.
+- `payload: Object` : optional parameters that the client specifies in `connectionParams`
+
+#### GQL_CONNECTION_TERMINATE
+Client sends this message to terminate the connection.
+
+#### GQL_START
+Client sends this message to execute GraphQL operation
+- `query: string` : GraphQL operation as string or parsed GraphQL document node
+- `variables?: Object` : Object with GraphQL variables
+- `operationName?: string` : GraphQL operation name
+- `context?: any` : Execution context for the operation
+    
+#### GQL_STOP
+Client sends this message in order to stop a running GraphQL operation execution (for example: unsubscribe)
+- `id: string` : operation id
+    
+### Server -> Client
+
+#### GQL_CONNECTION_ERROR
+The server sends this message if `onConnect` callback returns `false` or throws an exception. after sending this message, the server disconnects the client.
+- `payload: Object`: the server side error
+
+#### GQL_CONNECTION_ACK
+The server sends this message if `onConnect` callback returns any other value then `false`.
+
+#### GQL_DATA
+The server sends this message to transfter the GraphQL execution result from the server to the client.
+- `id: string` : ID of the operation that was successfully set up
+- `payload: { data: any }` : Execution result
```

Comment:
Why not present this as:

- `id: string`: ...
- `payload: object`
  -- `data: any`: ...
  -- `errors?: []` : ...